File: src/main/java/com/zippeykeys/praisebe/deity/Deity.java
package com.zippeykeys.praisebe.deity;

import com.google.common.collect.ImmutableMap;
import com.zippeykeys.praisebe.pattern.ILocalize;
import com.zippeykeys.praisebe.util.Util;

import org.immutables.value.Value.Immutable;
import org.immutables.value.Value.Style;
import org.immutables.value.Value.Style.ImplementationVisibility;
import org.jetbrains.annotations.Contract;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.ToString;
import lombok.val;
import net.minecraftforge.registries.IForgeRegistryEntry;

@Style(visibility = ImplementationVisibility.PACKAGE, overshadowImplementation = true, depluralize = true, strictBuilder = true)

@Immutable(copy = false)
public abstract class Deity extends IForgeRegistryEntry.Impl<Deity> implements ILocalize {
    @Override
    public abstract String getName();

    public abstract Type getType();

    public abstract Element getElement();

    public abstract Alignment getAlignment();

    public abstract ImmutableMap<String, Relationship> getRelationships();

    @Override
    @Contract(pure = true)
    public String getPrefix() {
        return "deity";
    }

    public Relationship getRelationship(Deity deity) {
        return getRelationship(deity.getName());
    }

    public Relationship getRelationship(String deity) {
        return getRelationships().get(deity);
    }

    @Contract(" -> new")
    public static Builder builder() {
        return new Builder();
    }

    public static class Builder extends ImmutableDeity.Builder {
        @Override
        public Deity build() {
            val result = super.build();
            result.setRegistryName(Util.getResource(result.getName()));
            return result;
        }
    }

    @ToString
    @AllArgsConstructor
    public enum Type implements ILocalize {
        TERRESTRIAL("terrestrial"), //
        CELESTIAL("celestial"), //
        ETHEREAL("ethereal");

        @Getter(onMethod_ = @Override)
        private final String name;

        @Override
        @Contract(pure = true)
        public String getPrefix() {
            return "deity.type";
        }
    }

    @ToString
    @AllArgsConstructor
    public enum Element implements ILocalize {
        AIR("air"), //
        EARTH("earth"), //
        FIRE("fire"), //
        WATER("water");

        @Getter(onMethod_ = @Override)
        private final String name;

        @Override
        @Contract(pure = true)
        public String getPrefix() {
            return "deity.element";
        }
    }

    @ToString
    @AllArgsConstructor
    public enum Alignment implements ILocalize {
        GOOD("good"), //
        NEUTRAL("neutral"), //
        EVIL("evil");

        @Getter(onMethod_ = @Override)
        private final String name;

        @Override
        @Contract(pure = true)
        public String getPrefix() {
            return "deity.alignment";
        }
    }

    @ToString
    @AllArgsConstructor
    public enum Relationship implements ILocalize {
        HATED("hated"), //
        DISLIKED("disliked"), //
        INDIFFERENT("indifferent"), //
        LIKED("liked"), //
        LOVED("loved"), //
        SELF("self");

        @Getter(onMethod_ = @Override)
        private final String name;

        @Override
        @Contract(pure = true)
        public String getPrefix() {
            return "deity.relationship";
        }
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/deity/Deity.kt
package com.zippeykeys.praisebe.deity

import com.google.common.collect.ImmutableMap
import com.zippeykeys.praisebe.iface.ILocalize
import net.minecraftforge.registries.IForgeRegistryEntry

data class Deity constructor(private val name: String, val type: Type, val element: Element, val alignment: Alignment, val relationships: ImmutableMap<String, Relationship>) : IForgeRegistryEntry.Impl<Deity>(), ILocalize {
    override val prefix: String
        get() = "deity"

    override fun getName(): String {
        return name
    }

    fun getRelationship(deity: Deity): Relationship? {
        return getRelationship(deity.name)
    }

    fun getRelationship(deity: String): Relationship? {
        return relationships[deity]
    }

    enum class Type constructor(private val actualName: String) : ILocalize {
        TERRESTRIAL("terrestrial"),
        CELESTIAL("celestial"),
        ETHEREAL("ethereal");

        override val prefix: String
            get() = "deity.type"

        override fun getName(): String {
            return actualName
        }
    }

    enum class Element constructor(private val actualName: String) : ILocalize {
        AIR("air"),
        EARTH("earth"),
        FIRE("fire"),
        WATER("water");

        override val prefix: String
            get() = "deity.element"

        override fun getName(): String {
            return actualName
        }
    }

    enum class Alignment constructor(private val actualName: String) : ILocalize {
        GOOD("good"),
        NEUTRAL("neutral"),
        EVIL("evil");

        override val prefix: String
            get() = "deity.alignment"

        override fun getName(): String {
            return actualName
        }
    }

    enum class Relationship constructor(private val actualName: String) : ILocalize {
        HATED("hated"),
        DISLIKED("disliked"),
        INDIFFERENT("indifferent"),
        LIKED("liked"),
        LOVED("loved"),
        SELF("self");

        override val prefix: String
            get() = "deity.relationship"

        override fun getName(): String {
            return actualName
        }
    }
}